new getPiece function

Canonelis opened the pull request:
Fixes the error at line 998 "getPieceId(object)" where an integer is sent to getPieceId.
Fixes the specific error described by https://steamcommunity.com/workshop/filedetails/discussion/2157495951/2993165479023391663/

Canonelis committed:
new getPiece function

Fixes the error at line 998 getPieceId(object) where an integer is sent to getPieceId.
Fixes the specific error described by https://steamcommunity.com/workshop/filedetails/discussion/2157495951/2993165479023391663/

Canonelis committed:
added debugging levels

Instead of debugging being on or off I included debugging levels and a button that cycles through them.
      -- 0: debugging off
      -- 1: specific function debugging
      -- 2: infrequent function calls
      -- 3: include frequent function calls
      -- 4: display everything

It allows contributors to control the rate of debug messages sent to the Atom console so they can see what is going on.  Level 1 is practically the same as filtering by the function name, so no debug messages are set at level 1 right now.

Also added a printTable function useful for debugging.  It will print the contents of tables and any other types of values.

Canonelis committed:
small check for getPieceId

Checks if object passed to getPieceId is userdata(ie. an object).

Canonelis committed:
changing the structure of the puzzle

Keep two variables for a group of joined pieces.
"parentId":The pieceId of the parent piece that is still on the board.
"joinedTo":a circularly linked list of all the pieces that make up the group.

eg. if two pieces 45 and 87 are connected then
gameState.pieces[45]={joinedTo:87,parentId=45,player_name="jack"}
and
gameState.pieces[87]={joinedTo:45,parentId=45,player_name="jack"}

Since the "joinedTo" value points to the id of the next piece in the chain, these values can easily be corrupted, but I took care not to mess up the code for this.

New conventions, if the piece is unjoined, then parentId == joinedTo == nil. Also the parent id of the parent piece is its own id, instead of nil.

This implementation includes an iterator "iterJoinedTo". It takes the piece id of any piece and iterates through its own id and all the ids of the attached pieces.
for id in iterJoinedTo(45) do
print(id)
end
-- output is "45" then "87" then stops
To exclude itself, you can use iterJoinedTo(45,false) and it will only output "87" and stop.

The JSON save is updated.

It also assigns a name to both target piece and dropped piece if the name is blank, so that it will ping all pieces without any gaps.

brainsick commented:
Do you have steps to recreate the linked error?  Any way to validate this fix?

Canonelis commented:
I didn't come up with a way to reproduce it 100%, but it was happening to me on numerous occasions too.
Tracing the code from line 1203 in the *.ttslua
"local targetPiece = getPiece(targetId)"
you can see how getPiece can't handle an id it has never seen before appropriately since it's a number.
The only validation right now is that this was the first fix I made and it's undergone over 60 hours of testing on TTS and it hasn't produced an error there anymone.

Canonelis committed:
made button that fixes deleted pieces and non-joining pices

This algorithm has a record of all valid GUIDs the puzzle started with.  If it finds any of the GUIDs are missing it respawns them.  Also, if a piece is not part of the initial valid GUIDs, then it will detach and delete it. However if the unknow GUID is a parent of many pieces (> 10), then keep it and make it a valid GUID.

Canonelis committed:
remove Wait.frames calls and use a callback function

use spawnObject()'s built-in callback function to assign a guid to guidCatalog once the object is done spawning.

brainsick commented:
I'm not following the basis of this change.  getPiece(object) is already designed to accept numbers.  That's the foundation of the neighbors/candidateMatches in onPieceDropped.

Canonelis committed:
set debugging level to default when game is loaded from a saved state

set debugging level to default when game is loaded from a saved state

Canonelis committed:
additional changes for fixing non-joining and deleted pieces

additional changes for fixing non-joining and deleted pieces
Producing debugging information for recovering broken puzzles
